### PR TITLE
Improve basis translator error message

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

When running the transpiler and the BasisTranslator is unable to translate the gates in a circuit to the specified target a moderately opaque error message is returned. In the context of the debugging the BasisTranslator pass it provides a bit more context but for the typical user encountering the error message it is a bit confusing because there is a bunch of superflous information that isn't actionable, and it also doesn't adequently explain the context of the error. This commit updates this error message to simplify the details returned so that it is just a list of the circuit's gates at the point of the translation and the specified target basis printed and also explaining why this error might be occuring. It also includes a link to the documentation on how to add custom equivalence rules just in case the target is using non-standard gates. This should hopefully make it clearer when transpile() fails in the BasisTranslator why this is occuring.

### Details and comments

### CODE REVIEW

**Review by jakelishman — APPROVED** [review]
I completely support doing this.  I'm just thinking about the cases when the basis translator runs as part of the optimisation loop - the gates the translator gets won't be the ones in the original circuit, so it might still be confusing in that situation.

That's pre-existing though, so it's not such a big deal if we can't come up with anything for this PR - it's already a big improvement over the old "could not map source basis <x> to target basis <y> over equivalence library <pointer>".

**Comment by jakelishman** [comment]
We have the general FAQ (though I'm not certain any of those question are frequently asked any more lol), but having a section in Terra docs that error messages can link to could possibly help?

**Comment by jakelishman** [comment]
There was still open discussion on this PR that wanted finalising before being added to a merge queue.

**Comment by mtreinish** [comment]
I added a small docs section to the BasisTranslator class documentation and updated the link in the error message to link to that instead.

File: qiskit/transpiler/passes/basis/basis_translator.py [hunk 1]
```diff
@@ -80,6 +80,24 @@ class BasisTranslator(TransformationPass):
         translating a 2 qubit operation on qubit 0 and 1 that the output might
         have ``u`` on qubit 1 and ``x`` on qubit 0. Typically running this pass
         a second time will correct these issues.
+
+    .. translation_errors:
```
Comment (jakelishman) [inline qiskit/transpiler/passes/basis/basis_translator.py:84]: I think you need a leading underscore on this link target?

Comment (mtreinish) [reply qiskit/transpiler/passes/basis/basis_translator.py:84]: Sigh, yeah I missed that. I should have run a docs build locally before pushing :)

**Review by jakelishman — APPROVED** [review]

**Pull request merged**